src/server/scene/basic_idle_hub.cpp

```diff
@@ -261,3 +275,18 @@ void ms::BasicIdleHub::schedule_alarm(ProofOfMutexLock const&, time::Timestamp c
         alarm_timeout = std::nullopt;
     }
 }
+
+std::shared_ptr<ms::IdleHub::WakeLock> ms::BasicIdleHub::inhibit_idle()
+{
+    if (wake_lock.expired()) // wake_lock is null
+    {
+        auto shared_lock = std::make_shared<BasicIdleHub::WakeLock>(shared_from_this());
+        wake_lock = shared_lock;  // WHERE SHOULD I GO?
+        alarm->cancel(); // TODO - move this logic into state logic
+        return shared_lock;
+    }
+    else // wake_lock is not null
+    {
+        return wake_lock.lock();
+    }
```

Comment:
This is racy. From [cppreference](https://en.cppreference.com/w/cpp/memory/weak_ptr/expired):
> This function is inherently racy if the managed object is shared among threads. In particular, a false result may become stale before it can be used. A true result is reliable. 

The object may not be expired when `.expired()` is checked, but is expired by the time `wake_lock.lock()` returns. This would lead to returning a null shared pointer. In this case the wake lock object isn't called into so it being null would just cause the wake lock not to work, but in many other cases it would lead to a crash.

Instead what you want to do is lock it first and then check if the returned shared pointer is null or not. If you put the lock in the if you can condense the code a bit:
```cpp
if (auto const shared_wake_lock = wake_lock.lock())
{
    return shared_wake_lock;
}
else
{
    ...
}
```

You also probably want to do any changes to the `BasicIdleHub` state (such as assigning the wake_lock and canceling the alarm) under lock.